### PR TITLE
feat: Added playground and preview feature

### LINKED ISSUE #2645: [FEATURE] Profile playground and check page
### Description

Playground page where people can add their profile json into a form and it shows them what it would look like

- validates json
- formats json
- shows main profile (bio, tags, social shortcuts)
- shows links
- shows milestones

### Screenshots

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by yatharth1706**
## Fixes Issue #2645 

closes #2645 

## Changes proposed

1. Added playground page
2. Added preview page

## Note to reviewers
Created a draft pull request. @eddiejaoude once review and let me know what you think.

<a href="https://gitpod.io/#https://github.com/EddieHubCommunity/LinkFree/pull/4211"></a>

### CODE REVIEW

File: components/user/UserProfile.js [hunk 1]
```diff
@@ -11,6 +11,7 @@ import { abbreviateNumber } from "../../services/utils/abbreviateNumbers";
 export default function UserProfile({ BASE_URL, data }) {
   const [qrShow, setQrShow] = useState(false);
   const fallbackImageSize = 120;
+  console.log(data);
```
Comment (eddiejaoude) [inline components/user/UserProfile.js:14]: I think this debug code can be removed

File: components/PreviewModal.js [hunk 2]
```diff
@@ -0,0 +1,187 @@
+import { useState, useEffect } from "react";
+import Link from "../components/Link";
+import { IconContext } from "react-icons";
+import { FaRegComments } from "react-icons/fa";
+import requestIp from "request-ip";
```
Comment (eddiejaoude) [inline components/PreviewModal.js:5]: Are all these imports used?

File: components/PreviewModal.js [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    </>
+  );
+}
+
+// User.getLayout = function getLayout(page) {
+//   if (singleUser.username) {
+//     return <SingleLayout>{page}</SingleLayout>;
+//   }
+//   return <MultiLayout>{page}</MultiLayout>;
+// };
```
Comment (eddiejaoude) [inline components/PreviewModal.js:187]: I think commented out code can be removed?

File: pages/playground.js [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        setErrMsg("");
+      }
+    } catch (err) {
+      setErrMsg(err.toString());
+      setSuccessMsg("");
+    }
+  };
+
+  const handleReset = () => {
+    localStorage.removeItem("PREVIEW_PROFILE_DATA");
```
Comment (eddiejaoude) [inline pages/playground.js:65]: Is local storage still needed if used in a modal on the same page?

File: components/PreviewModal.js [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+  const BASE_URL = process.env.NEXT_PUBLIC_BASE_URL;
+
+  useEffect(() => {
+    if (isOpen) {
+      handlePreviewData();
+    }
+  }, [isOpen]);
+
+  const handlePreviewData = () => {
+    let userPreviewData = localStorage.getItem("PREVIEW_PROFILE_DATA");
```
Comment (eddiejaoude) [inline components/PreviewModal.js:29]: I think this could be passed into the component as a `data` object

File: components/PreviewModal.js [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    let parsedData = JSON.parse(userPreviewData);
+    setUserData(parsedData);
+    console.log(parsedData);
```
Comment (eddiejaoude) [inline components/PreviewModal.js:32]: I think the debug code can be removed?

Comment (yatharth1706) [reply components/PreviewModal.js:32]: Thanks @eddiejaoude for reviewing the code. Will make all these changes which you suggested

**Comment by yatharth1706** [comment]
Hy @eddiejaoude, I have done all the changes which you suggested
- Removed local storage code
- Removed debug code as well like console.log statements
- Removed unused libraries
- Removed commented code
- Passing data prop to preview instead of taking data from local storage

**Comment by eddiejaoude** [comment]
Looks good. Just had a quick play and when navigating to Testimonial tab I get an error

Screenshot 2023-02-08 at 07 26 14

**Comment by yatharth1706** [comment]
> Looks good. Just had a quick play and when navigating to Testimonial tab I get an error
> 
> Screenshot 2023-02-08 at 07 26 14

Ok @eddiejaoude, I will take a look at this why it is happening. Will solve it by EOD or tomorrow

**Comment by yatharth1706** [comment]
Hy @eddiejaoude testimonials one can be little tricky, I think. 

I was checking your code and whenever user go to /[username].  API is called which gets all the json and testimonials from data folder and combine all testimonials like its title, description, date in the json.

For my playground, I am not making any api calls. I am using the json which user provides in the UI. And according to that UI showing the preview. And for testimonials, I think will have to ask more information in the json. As our playground is totally dependent on json. Its not going to testimonials folder to find full info of testimonials given by particular user

Need your ideas as well. I think for now our playground will not work properly with testimonials. Either we should change the format of json and provide full information of testimonials. Currently in testimonials only gitusername is written. If we write full info like title, description, date along with git usernames. Then i can make changes accordingly to show that in preview.

**Comment by eddiejaoude** [comment]
Thank @yatharth1706 for investigating 👍 I have some ideas, and I think the best solution would be to break the profile  page into smaller reusable components then you can use the components you want for your feature - I agree the testimonials are not needed for this, keep this feature simple would be best

The challenge though is doing this refactor of the app is a lot of work and will cause conflicts - plus I think should be in a separate PR to this one - I have started doing it but still a long way to go.

So maybe for now this PR is blocked by another PR and needs to be parked

**Comment by yatharth1706** [comment]
Hy @eddiejaoude, I actually didnt understand properly. So the other PR which you mentioned, are you working on it ?

**Comment by eddiejaoude** [comment]
> Hy @eddiejaoude, I actually didnt understand properly. So the other PR which you mentioned, are you working on it ?

Sorry if my message was not clear. Yes I have started breaking it up into reusable components. I will see if there is a short term solution we can use to get this PR merged sooner

**Comment by yatharth1706** [comment]
Ok @eddiejaoude

**Comment by eddiejaoude** [comment]
A recently merged PR might help this PR, I will take a look https://github.com/EddieHubCommunity/LinkFree/pull/4997/files#diff-45322352e129989f2db77f2dffd9308e2420ba561efa94d938cadb79fb27c5e4

**Pull request merged**